text/0000-app-image-extensions.md

```diff
@@ -0,0 +1,117 @@
+# Meta
+[meta]: #meta
+- Name: App Image Extensions
+- Start Date: 2019-08-09
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes:
+1. An interface between a set of stack images and a platform in order to extend the images
+2. A UX for the pack CLI for extending builders with OS packages
+
+# Motivation
+[motivation]: #motivation
+
+Allowing buildpacks to install OS packages dynamically during build would drastically increase build time, especially when CNB tooling is used to build or rebase many apps with similar package requirements.
+
+Mixins already allow buildpack authors to create buildpacks that depend on an extended set of OS packages without affecting build time.
+However, it is not uncommon for application code to depend on OS packages.
+
+Given that app authors and platform maintainers experience increased build time directly, extending the stack at the app author's or platform maintainer's request may add flexibility without sacrificing performance.
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Specification
+
+Stack image creators may add the following executables to their run and/or build images:
+
+```
+/cnb/image/build/extend (pkg-cache-dir) (build-metadata-toml-file) | cwd: /
+/cnb/image/run/extend (pkg-cache-dir) (run-metadata-toml-file) | cwd: /
+```
+* executes on base run or build images
+* writes metadata to `/cnb/image/build/metadata.toml` or `/cnb/image/run/metadata.toml`
+* on subsequent builds, lookup latest upstream base image digest:
+  * if the digest is different than extended image's base, the new upstream digest is extended
+  * if the digest is the same, proceed to status
+
+```
+/cnb/image/build/status (pkg-cache-dir) (build-metadata-toml-file) | cwd: /
+/cnb/image/run/status (pkg-cache-dir) (run-metadata-toml-file) | cwd: /
+```
+* executes on run or build base images
+* metadata arg is `/cnb/image/build/metadata.toml` or `/cnb/image/run/metadata.toml`
+* exit status 100 = existing base image digest is extended 
+* exit status 0 = does not need extension
+* exit status other = unknown
+
+The status executable may be invoked to determine whether updates are necessary.
+
+## UX
+
+### Creating an Extended Builder
+
+`pack create-builder sclevine/builder -b builder.toml`
```

Comment:
I did a GitHub search, but there isn't a spec around what a `builder.toml` is. Is there a single place where I can read what defines that/how it's meant to be used?

Reply:
As well, would this be considered to be idempotent? As in, multiple invocations of the same `create-builder` command with the same args should be a noop?

Reply:
The spec you're looking for is here: https://buildpacks.io/docs/using-pack/working-with-builders/#builder-configuration

The output image is always re-created from images specified in builder.toml, so I'm not sure what idempotent would mean in this context. Given the same inputs, the output should always be the same. It's not an operation that can be re-applied. (And `extend` is never run on an image that is already extended.)

Reply:
I mean like, if I wanted to put this in the context of my own platform, could I always run `create-builder` with that toml in place and expect it to return immediately because the input `builder.toml` was the same? That way I don't have to wait for an expensive rebuild step on each "deploy" of my app.

Reply:
`pack create-builder` could check to see if there's an existing builder at the specified tag, verify that its metadata matches builder.toml, swap out any buildpacks, and run `pack upgrade` on it.

Similarly, if the extended builder already exists, `pack extend-builder` could swap out buildpacks with those in the source image and run `pack upgrade`.

As long as the base image digests haven't changed and `status` passes, you wouldn't need to wait for a rebuild. `pack upgrade` will check this for you.

Reply:
If the answer turns from _could_ to _does_, I can see that being useful to many users, as otherwise there wouldn't be much of a difference from an image extension to a buildpack that vendors installed packages to a subdirectory in the workspace.